contributors/design-proposals/kubeadm-upgrades.md

```diff
@@ -0,0 +1,452 @@
+
+ Upgrade to the latest patch version of your branch:
+
+			Current version	Upgrade available
+
+ API server:		v1.7.5		v1.7.8
+
+ Controller-manager:	v1.7.5		v1.7.8
+
+ Scheduler:		v1.7.5		v1.7.8
+
+ Kube-proxy:		v1.7.5		v1.7.8
+
+ Kube-dns:			v1.14.2		v1.14.4
+
+ Etcd:			v3.0.17		v3.1.10
+
+ You can now apply the upgrade by executing the following command:
+
+	kubeadm upgrade apply --version v1.7.8
+
+Upgrade to the latest stable minor release:
+
+			Current version	Upgrade available
+
+ API server:		v1.7.5		v1.8.3
+
+ Controller-manager:	v1.7.5		v1.8.3
+
+ Scheduler:		v1.7.5		v1.8.3
+
+ Kube-proxy:		v1.7.5		v1.8.3
+
+ Kube-dns:			v1.14.2		v1.15.5
+
+ You can now apply the upgrade by executing the following command:
+
+	kubeadm upgrade apply --version v1.8.3
+
+ Components that must be upgraded manually:
+
+ Kubelet			2 x v1.8.0		v1.8.3
+
+				3 x v1.8.1		v1.8.3
+
+# Implementation options of the upgrade
+
+High-level approaches:
+
+1. Build the upgrade logic into kubeadm
+
+2. Build an upgrading Operator that does the upgrading for us
+
+    * Would consume a TPR with details how to do the upgrade
+
+The historical preference was to build a reusable Operator that does the upgrading for us. *Either way we would deliver the same 'kubeadm upgrade' UX.* In the latter case, that would involve kubeadm deploying an upgrading controller on behalf of the user, and tracking its lifecycle through the kubernetes API.
+
+There are pros and cons for each method:
+
+1. Build it into the `kubeadm` CLI command
+
+    1. Pros:
+
+        1. *Simplicity*; Very straightforward to implement; basically doing the same steps as `kubeadm init` would, only slightly modified to fit the upgrading purpose. We will construct a `kubeadm upgrade` command in any case
+
+        2. The kubeadm v1.6 -> v1.7 upgrade used this method this successfully
+
+        3. Being able to easily have an interactive flow where the user can inspect the changes that are about to be applied and accept/deny for instance.
+
+        4. Doesn’t have to deal with any version skew between the Operator image tag/version and the CLI version
+
+        5. Doesn’t have to define an API (CRD or the like) between the client and the Operator
+
+    2. Cons:
+
+        6. Maybe not as reusable as an Operator inside the cluster would be
+
+2. Build an Operator
+
+    3. Pros:
+
+        7. The idea is that it would be generic and could be re-used by others
+
+        8. Other operators are already handling complex upgrades successfully, and we could even standardise patterns
+
+    4. Cons:
+
+        9. Has to deal with any version skew between the controller and the CLI
+
+        10. Has to deal with pre-pulling the image to a specific node in the case of no internet connection
+
+        11. Has to define an API (CRD or the like) between the client and the Operator
+
+**Decision**: Keep the logic inside of the kubeadm CLI (option 1) for the implementation in v1.8.0.
```

Comment:
This makes sense. I don't think it would be difficult to move the logic later (at least not more difficult than using an operator now).